src/BizHawk.Emulation.Cores/Consoles/3DO/Opera.cs

```diff
@@ -0,0 +1,320 @@
+		private static int CD_SECTOR_SIZE = 2048;
+		private readonly byte[] _sectorBuffer = new byte[CD_SECTOR_SIZE];
+
+		private void SelectNextDisc()
+		{
+			_discIndex++;
+			if (_discIndex == _discAssets.Count) _discIndex = 0;
+			CoreComm.Notify($"Selected CDROM {_discIndex}: {_discAssets[_discIndex].DiscName}", null);
+		}
+
+		private void SelectPrevDisc()
+		{
+			_discIndex--;
+			if (_discIndex < 0) _discIndex = _discAssets.Count - 1;
+			CoreComm.Notify($"Selected CDROM {_discIndex}: {_discAssets[_discIndex].DiscName}", null);
+		}
+
+		private void CDRead(int lba, IntPtr dest)
+		{
+			if (_discIndex < _discAssets.Count)
+			{
+				_cdReaders[_discIndex].ReadLBA_2048(lba, _sectorBuffer, 0);
+				Marshal.Copy(_sectorBuffer, 0, dest, CD_SECTOR_SIZE);
+			}
+			DriveLightOn = true;
+		}
+
+		private int CDSectorCount()
+		{
+			if (_discIndex < _discAssets.Count) return _discAssets[_discIndex].DiscData.Session1.LeadoutLBA;
+			return -1;
+		}
+
+		protected override LibWaterboxCore.FrameInfo FrameAdvancePrep(IController controller, bool render, bool rendersound)
+		{
+			var fi = new LibOpera.FrameInfo();
+
+			// Disc management
+			if (_isMultidisc)
+			{
+				if (controller.IsPressed("Next Disc")) SelectNextDisc();
+				if (controller.IsPressed("Prev Disc")) SelectPrevDisc();
+			}
+
+			DriveLightOn = false;
+			fi.port1 = ProcessController(1, _syncSettings.Controller1Type, controller);
+			fi.port2 = ProcessController(2, _syncSettings.Controller2Type, controller);
+
+			// Game reset
+			if (controller.IsPressed("Reset")) fi.isReset = 1;
+
+			return fi;
+		}
+
+		private static LibOpera.GameInput ProcessController(int port, ControllerType type, IController controller)
+		{
+			LibOpera.GameInput gameInput = new LibOpera.GameInput();
+
+			switch (type)
+			{
+				case ControllerType.Gamepad:
+					gameInput.gamepad.up = controller.IsPressed($"P{port} {JoystickButtons.Up}") ? 1 : 0;
+					gameInput.gamepad.down = controller.IsPressed($"P{port} {JoystickButtons.Down}") ? 1 : 0;
+					gameInput.gamepad.left = controller.IsPressed($"P{port} {JoystickButtons.Left}") ? 1 : 0;
+					gameInput.gamepad.right = controller.IsPressed($"P{port} {JoystickButtons.Right}") ? 1 : 0;
+					gameInput.gamepad.buttonX = controller.IsPressed($"P{port} {JoystickButtons.ButtonX}") ? 1 : 0;
+					gameInput.gamepad.buttonP = controller.IsPressed($"P{port} {JoystickButtons.ButtonP}") ? 1 : 0;
+					gameInput.gamepad.buttonA = controller.IsPressed($"P{port} {JoystickButtons.ButtonA}") ? 1 : 0;
+					gameInput.gamepad.buttonB = controller.IsPressed($"P{port} {JoystickButtons.ButtonB}") ? 1 : 0;
+					gameInput.gamepad.buttonC = controller.IsPressed($"P{port} {JoystickButtons.ButtonC}") ? 1 : 0;
+					gameInput.gamepad.buttonL = controller.IsPressed($"P{port} {JoystickButtons.ButtonL}") ? 1 : 0;
+					gameInput.gamepad.buttonR = controller.IsPressed($"P{port} {JoystickButtons.ButtonR}") ? 1 : 0;

```

Comment:
I feel like this method can be cleaned up further

Reply:
Any suggestions?

Reply:
Maybe:
```csharp
var pfx = $"P{port} ";
int IsPressed(string button) => controller.IsPressed(pfx + button) ? 1 : 0;
gameInput.gamepad.buttonB = IsPressed(JoystickButtons.ButtonB);
```

Reply:
I see the point being made, but personally I value readability and maintainability pretty highly. Even if this comes at the price of some repetition.

The proposed change doesn't reduce the number of LoC, but in fact adds an indirection for the meaning of pfx making, imo, the code ever so slightly harder to read.

I'm not opposed to the change if the maintainers think it's better, but I'd personally leave this unchanged.

Reply:
Indeed, doesn't look too much simpler than the original. Doesn't feel like a qualitative improvement, but rather a matter of style, so I'd leave it as is I guess.